wal/wal.go

```diff
@@ -0,0 +1,825 @@
+func (w *WAL) log(rec []byte, final bool) error {
+	w.mtx.Lock()
+	defer w.mtx.Unlock()
+
+	// If the record is too big to fit within pages in the current
+	// segment, terminate the active segment and advance to the next one.
+	// This ensures that records do not cross segment boundaries.
+	left := w.page.remaining() - recordHeaderSize                                   // active page
+	left += (pageSize - recordHeaderSize) * (w.pagesPerSegment() - w.donePages - 1) // free pages
+
+	if len(rec) > left {
+		if err := w.nextSegment(); err != nil {
+			return err
+		}
+	}
+
+	// Populate as many pages as necessary to fit the record.
+	// Be careful to always do one pass to ensure we write zero-length records.
+	for i := 0; i == 0 || len(rec) > 0; i++ {
+		p := w.page
+
+		// Find how much of the record we can fit into the page.
+		var (
+			l    = min(len(rec), (pageSize-p.alloc)-recordHeaderSize)
+			part = rec[:l]
+			buf  = p.buf[p.alloc:]
+			typ  recType
+		)
+
+		switch {
+		case i == 0 && len(part) == len(rec):
+			typ = recFull
+		case len(part) == len(rec):
+			typ = recLast
+		case i == 0:
+			typ = recFirst
+		default:
+			typ = recMiddle
+		}
+
+		buf[0] = byte(typ)
+		crc := crc32.Checksum(part, castagnoliTable)
+		binary.BigEndian.PutUint16(buf[1:], uint16(len(part)))
+		binary.BigEndian.PutUint32(buf[3:], crc)
+
+		copy(buf[7:], part)
+		p.alloc += len(part) + recordHeaderSize
+
+		// If we wrote a full record, we can fit more records of the batch
+		// into the page before flushing it.
+		if final || typ != recFull || w.page.full() {
+			if err := w.flushPage(false); err != nil {
+				return err
+			}
+		}
+		rec = rec[l:]
+	}
+	return nil
+}
+
+// Segments returns the range [m, n] of currently existing segments.
+// If no segments are found, m and n are -1.
+func (w *WAL) Segments() (m, n int, err error) {
+	refs, err := listSegments(w.dir)
+	if err != nil {
+		return 0, 0, err
+	}
+	if len(refs) == 0 {
+		return -1, -1, nil
+	}
+	return refs[0].n, refs[len(refs)-1].n, nil
+}
+
+// Truncate drops all segments before i.
+func (w *WAL) Truncate(i int) error {
+	refs, err := listSegments(w.dir)
+	if err != nil {
+		return err
+	}
+	for _, r := range refs {
+		if r.n >= i {
+			break
+		}
+		if err := os.Remove(filepath.Join(w.dir, r.s)); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (w *WAL) fsync(f *Segment) error {
+	start := time.Now()
+	err := fileutil.Fsync(f.File)
+	w.fsyncDuration.Observe(time.Since(start).Seconds())
+	return err
+}
+
+// Close flushes all writes and closes active segment.
+func (w *WAL) Close() (err error) {
+	w.mtx.Lock()
+	defer w.mtx.Unlock()
+
+	// Flush the last page and zero out all its remaining size.
+	// We must not flush an empty page as it would falsely signal
+	// the segment is done if we start writing to it again after opening.
+	if w.page.alloc > 0 {
+		if err := w.flushPage(true); err != nil {
+			return err
+		}
+	}
+
+	donec := make(chan struct{})
```

Comment:
I think a single-level channel is sufficient here

Reply:
It's not, we have to wait until run terminated successfully. Currently it returns right when it reads from the channel. But this is quite implementation dependent and would be an easy regression target.

Reply:
It looks like unnecessary complexity currently, especially as WAL is a struct rather than an interface.